File: Python/union-find-sets/path_with_minimum_effort.py
import collections
import heapq


def minimumEffortPath_binary_bfs(heights: list) -> int:
    m, n = len(heights), len(heights[0])
    left, right, ans = 0, 10**6 - 1, 0

    while left <= right:
        mid = (left + right) // 2
        q = collections.deque([(0, 0)])
        seen = {(0, 0)}

        while q:
            x, y = q.popleft()
            for nx, ny in [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)]:
                if 0 <= nx < m and 0 <= ny < n and (nx, ny) not in seen and abs(heights[x][y] - heights[nx][ny]) <= mid:
                    q.append((nx, ny))
                    seen.add((nx, ny))

        if (m - 1, n - 1) in seen:
            ans = mid
            right = mid - 1
        else:
            left = mid + 1
    return ans


def dfs(seen: list, row: int, col: int, mid: int, row_len, col_len, heights):
        seen.append((row, col))
        for nx, ny in [(row - 1, col), (row + 1, col), (row, col + 1), (row, col - 1)]:
            if 0 <= nx < row_len and 0 <= ny < col_len and abs(heights[row][col] - heights[nx][ny]) <= mid and (nx, ny) not in seen:
                seen.append((nx, ny))
                dfs(seen, nx, ny, mid, row_len, col_len, heights)

def minimumEffortPath_binary_dfs(heights: list) -> int:
    row_len, col_len = len(heights), len(heights[0])
    left, right, result = 0, 10**6 - 1, 0
    while left <= right:
        seen = []
        mid = (left + right) // 2
        dfs(seen, 0, 0, mid, row_len, col_len, heights)

        if (row_len - 1, col_len - 1) in seen:
            result = mid
            right = mid - 1
        else:
            left = mid + 1
        
        print(seen)

    return result

def find(parent: list, index: int) -> int:
    if index != parent[index]:
        parent[index] = find(parent, parent[index])
    return parent[index]

def union(parent: list, index1: int, index2: int) -> None:
    parent[find(parent, index2)] = find(parent, index1)

def minimumEffortPath_union_find(heights: list) -> int:
    row, col = len(heights), len(heights[0])
    parent = list(range(row * col))

    edges = []
    for i in range(row):
        for j in range(col):
            if i != row - 1:
                edges.append((abs(heights[i][j] - heights[i + 1][j]), i * col + j, (i + 1) * col + j))
            if j != col - 1:
                edges.append((abs(heights[i][j] - heights[i][j + 1]), i * col + j, i * col + j + 1))
    
    result = 0
    for weight, x, y in sorted(edges):
        union(parent, x, y)
        result = weight
        if find(parent, 0) == find(parent, (row - 1) * col + (col - 1)):
            break
    
    return result

def minimumEffortPath_dijkstra(heights: list) -> int:
    m, n = len(heights), len(heights[0])
    q = [(0, 0, 0)]
    dist = [0] + [float("inf")] * (m * n - 1)
    seen = set()

    while q:
        d, x, y = heapq.heappop(q)
        iden = x * n + y
        if iden in seen:
            continue
        if (x, y) == (m - 1, n - 1):
            break

        seen.add(iden)
        for nx, ny in [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)]:
            if 0 <= nx < m and 0 <= ny < n and max(d, abs(heights[x][y] - heights[nx][ny])) <= dist[nx * n + ny]:
                dist[nx * n + ny] = max(d, abs(heights[x][y] - heights[nx][ny]))
                heapq.heappush(q, (dist[nx * n + ny], nx, ny))
    
    return dist[m * n - 1]


if __name__ == "__main__":
    print(minimumEffortPath_dijkstra([[4,3,4,10,5,5,9,2],[10,8,2,10,9,7,5,6],[5,8,10,10,10,7,4,2],[5,1,3,1,1,3,1,9],[6,4,10,6,10,9,4,6]]))

File: core/tmp/Java/src/com/dbc/algorithm/Route/PathWithMinimumEffort.java
package com.dbc.algorithm.Route;

import java.util.*;

public class PathWithMinimumEffort {
    public static int minimumEffortPathBfs(int[][] heights) {
        int left = 0, right = (int) Math.pow(10, 6) - 1, result = 0;
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int row = heights.length, col = heights[0].length;
        Queue<int[]> queue = new LinkedList<>();
        while (left <= right) {
            int[] seen = new int[row * col];
            int mid = (left + right) / 2;
            seen[0] = 1;
            queue.offer(new int[]{0, 0});
            while (!queue.isEmpty()) {
                int[] temp = queue.poll();
                for (int[] dir : dirs) {
                    int nx = temp[0] + dir[0];
                    int ny = temp[1] + dir[1];
                    if (nx >= 0 && nx < row && ny >= 0 && ny < col && seen[nx * col + ny] != 1 && Math.abs(heights[temp[0]][temp[1]] - heights[nx][ny]) <= mid) {
                        seen[nx * col + ny] = 1;
                        queue.offer(new int[]{nx, ny});
                    }
                }
            }

            if (seen[(row - 1) * col + col - 1] == 1) {
                right = mid - 1;
                result = mid;
            } else {
                left = mid + 1;
            }
        }

        return result;
    }

    public static void dfs(int[][] heights, int[] seen, int row, int col, int mid) {
        seen[row * heights[0].length + col] = 1;
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        for (int[] dir : dirs) {
            int nx = dir[0] + row;
            int ny = dir[1] + col;
            if (nx >= 0 && nx < heights.length && ny >= 0 && ny < heights[0].length && seen[nx * heights[0].length + ny] != 1 && Math.abs(heights[row][col] - heights[nx][ny]) <= mid) {
                seen[nx * heights[0].length + ny] = 1;
                dfs(heights, seen, nx, ny, mid);
            }
        }
    }

    public static int minimumEffortPathDfs(int[][] heights) {
        int left = 0, right = (int) Math.pow(10, 6) - 1, result = 0;
        int row = heights.length, col = heights[0].length;

        while (left <= right) {
            int mid = (left + right) / 2;
            int[] seen = new int[row * col];
            seen[0] = 1;
            dfs(heights, seen, 0, 0, mid);

            if (seen[(row - 1) * col + col - 1] == 1) {
                right = mid - 1;
                result = mid;
            } else {
                left = mid + 1;
            }
        }

        return result;
    }

    public static int find(int[] parent, int index) {
        if (index != parent[index]) {
            parent[index] = find(parent, parent[index]);
        }
        return parent[index];
    }

    public static void union(int[] parent, int index1, int index2) {
        parent[find(parent, index2)] = find(parent, index1);
    }

    public static int minimumEffortPathUnionFind(int[][] heights) {
        List<int[]> edges = new LinkedList<>();
        int row = heights.length, col = heights[0].length;
        int[] parent = new int[row * col];
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                parent[i * col + j] = i * col + j;
                if (i < row - 1) {
                    edges.add(new int[]{Math.abs(heights[i + 1][j] - heights[i][j]), i * col + j, (i + 1) * col + j});
                }
                if (j < col - 1) {
                    edges.add(new int[]{Math.abs(heights[i][j] - heights[i][j + 1]), i * col + j, i * col + j + 1});
                }
            }
        }

        Collections.sort(edges, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return o1[0] - o2[0];
            }
        });

        for (int[] edge : edges) {
            int rootX = find(parent, edge[1]);
            int rootY = find(parent, edge[2]);
            if (rootX != rootY) {
                union(parent, rootX, rootY);
            }
            if (find(parent, 0) == find(parent, (row - 1) * col + col - 1))
                return edge[0];
        }

        return 0;
    }

    public static int minimumEffortPathDijkstra(int[][] heights) {
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        int m = heights.length;
        int n = heights[0].length;
        PriorityQueue<int[]> pq = new PriorityQueue<>(new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return o1[2] - o2[2];
            }
        });

        pq.offer(new int[]{0, 0, 0});
        int[] dist = new int[m * n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[0] = 0;
        boolean[] seen = new boolean[m * n];
        while (!pq.isEmpty()) {
            int[] edge = pq.poll();
            int x = edge[0], y = edge[1], d = edge[2];
            int id = x * n + y;
            if (seen[id]) {
                continue;
            }
            seen[id] = true;
            for (int i = 0; i < 4; i++) {
                int nx = x + dirs[i][0];
                int ny = y + dirs[i][1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && Math.max(d, Math.abs(heights[nx][ny] - heights[x][y])) < dist[nx * n + ny]) {
                    dist[nx * n + ny] = Math.max(d, Math.abs(heights[x][y] - heights[nx][ny]));
                    pq.offer(new int[]{nx, ny, dist[nx * n + ny]});
                }
            }
        }
        return dist[m * n - 1];
    }
}
